### ISSUE TITLE
[IMPROVEMENT] Cannot expand a volume created by Longhorn UI

### ISSUE THREAD

**Issue by PhanLe1010** [longhorn/longhorn#6446]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

Let's say user manually creates Longhorn volume and PV/PVC, the PV/PVC will have longhorn-static StorageClass. This StorageClass does not actually exist. If the user wants to expand the volume, they will use one of the 2 following approaches:

1. Expand the volume using Longhorn UI: we try to expand the PVC first before attempting trying to expand the Longhorn volume. The [PVC expansion operation](https://github.com/longhorn/longhorn-manager/blob/e278ba0c2d2591dfae8f79b069574bf2ede7e330/manager/volume.go#L481-L482) always fails because the PVC is not dynamic since it is associated with longhorn-static StorageClass:
    ```
    [longhorn-manager-8pfb7] time="2023-08-01T18:36:49Z" level=warning msg="HTTP handling error" error="unable to expand volume testvol: persistentvolumeclaims \"testvol\" is forbidden: only dynamically provisioned pvc can be resized and the storageclass that provisions the pvc must support resize"
    ```
1. Expand the volume through the CSI flow by updating the PVC.Spec directly, this will of course fails with the same webhook error:
    ```
    persistentvolumeclaims "testvol" is forbidden: only dynamically provisioned pvc can be resized and the storageclass that provisions the pvc must support resize
    ```

## Describe the solution you'd like

@shuo-wu @ejweber @james-munson and me agreeed on this solution:

Create a real StorageClass with the name set in Default Longhorn Static StorageClass Name and only the bare minimum values so that expansion "just works". If that setting changes, create a new StorageClass with the new name. 

Write a KB about a workaround for users with PVC with `longhorn-static` SC in older version

## Alternatives we have considered

1. Same as above except change the name to longhorn-ui first if the setting has never been modified. This alleviates the concerns about confusion around "static", but has some added complexity.
1. Just change the default SC setting to `longhorn`. Now we don't have to create a new StorageClass to enable expansion, but there is the potential for confusion around why the parameters set in the longhorn StorageClass don't do anything.
1. Ask users to decide whether they need a "dynamic" volume or not in the UI and use either longhorn or the longhorn-static based on their response. In this situation, it's better just to take the "right" action automatically. Users won't know at volume creation whether they will want to expand the volume in the future and won't remember in the future when expansion is failing that they made this decision themselves.

**Comment by haoyixin** [comment]
+1

**Comment by SleepWalker** [comment]
Can anybody suggest a solution till expansion will work as expected?

I creating all volumes through UI because it's the only way to get custom VC/PVC names. Some of the volumes are encrypted and some RWX.

The only solution I've found so far:

- stop all pods
- delete PVC
- update PV declaration with new size
- resize PV in UI
- create a new PVC from UI (use previous PVC checkbox)

It's pain 😢

**Comment by PhanLe1010** [comment]
Workaround while waiting for the improvement:

Deploy this dummy storageclass then the expansion will work:
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: longhorn-static
provisioner: driver.longhorn.io
allowVolumeExpansion: true
parameters:
  staleReplicaTimeout: "30" # 30 minutes
```

**Comment by SleepWalker** [comment]
> Deploy this dummy storageclass then the expansion will work:

However in this case if you'll create PVC, longhorn will automatically provision corresponding PV with the random name (if I'm not mistaken). This will be the difference with default behavior. But this may be acceptable for the most users :thinking: (I have a weird setup for my homelab where I want to have meaningful names)

But in any case it's much more pain to resize everything manually with the downtime. So thank you very much for suggestion!

**Comment by PhanLe1010** [comment]
> However in this case if you'll create PVC, longhorn will automatically provision corresponding PV with the random name (if I'm not mistaken). This will be the difference with default behavior. But this may be acceptable for the most users 🤔 (I have a weird setup for my homelab where I want to have meaningful names)

No, after deploy the dummy storageclass, you can continue the old flow: create a volume using Longhorn UI -> Create PV/PVC with your chosen names using Longhorn UI -> Expand the volume using Longhorn UI. Everything should work

**Comment by tubbo** [comment]
I only have one storage class:

```
NAME                 PROVISIONER          RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
longhorn (default)   driver.longhorn.io   Delete          Immediate           true                   27d
```

And it is using `ext4` with `allowVolumeExpansion` enabled:

```
Name:            longhorn
IsDefaultClass:  Yes
Annotations:     longhorn.io/last-applied-configmap=kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: longhorn
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: driver.longhorn.io
allowVolumeExpansion: true
reclaimPolicy: "Delete"
volumeBindingMode: Immediate
parameters:
  numberOfReplicas: "1"
  staleReplicaTimeout: "30"
  fromBackup: ""
  fsType: "ext4"
  dataLocality: "disabled"
,storageclass.kubernetes.io/is-default-class=true
Provisioner:           driver.longhorn.io
Parameters:            dataLocality=disabled,fromBackup=,fsType=ext4,numberOfReplicas=1,staleReplicaTimeout=30
AllowVolumeExpansion:  True
MountOptions:          <none>
ReclaimPolicy:         Delete
VolumeBindingMode:     Immediate
Events:                <none>
```

Yet, Longhorn UI will still not allow me to expand the volume:

![Screenshot 2024-01-02 at 9 59 24 PM](https://github.com/longhorn/longhorn/assets/113026/8b053d96-2979-4ff6-9f6c-c83eea929a38)

I was able to change the PVC to my expected capacity, but Longhorn doesn't want to actually change the size of the volume.

**Comment by Floeig** [comment]
> Workaround while waiting for the improvement:
> 
> Deploy this dummy storageclass then the expansion will work:
> 
> ```
> kind: StorageClass
> apiVersion: storage.k8s.io/v1
> metadata:
>   name: longhorn-static
> provisioner: driver.longhorn.io
> allowVolumeExpansion: true
> parameters:
>   staleReplicaTimeout: "30" # 30 minutes
> ```

@PhanLe1010  Thanks a lot this solved my issue!!!  I'll still follow the thread if one day I shoud remove it. If I can be from any help let me know too

**Comment by chriscchien** [comment]
Verified **pass** on longhorn master(longhorn-manager `77ac26`)

1. Create volume `vol1` with size 1 Gi from UI
2. Create PVC for `vol1` from UI
3. Use a workload to mount `vol1`, wait pod running
4. Expand volume `vol1`  to 3 Gi from UI
5. In pod, write data greater then 1 Gi to the mount point
6. Observed the actual size of volume is bigger than 1 Gi (In v1.6.2, the actual size will be around 1 Gi which is the size before volume expand)
